### PR TITLE
Fix heatmap labels

### PR DESCRIPTION
**Pull request by George-Zipperlen**
Rotate the gene-symbol labels in the heatmap

### CODE REVIEW

**Comment by dhimmel** [comment]
@George-Zipperlen did you also update `heatmap.ipynb` (the notebook)? If so can you add that file in addition to `heatmap.py`?

**Comment by George-Zipperlen** [comment]
@dhimmel <https://github.com/dhimmel> Yes,

I produced the .py file with jupyter -nbconvert

I wasn't sure from the workflow whether to add the .ipynb file to the pull request.

Will do that now.

File: explore/heatmap-vega-lite/heatmap.py [hunk 1]
```diff
@@ -123,14 +124,14 @@
 y = y_full.assign(indicator = y_full.max(axis=1))
 
 
-# In[12]:
+# In[11]:
 
 unique_pos = y.groupby('disease').apply(lambda x: x['indicator'].sum())
 heatmap_df0 = y_full.groupby('disease').sum().assign(TOTAL = unique_pos)
```
Comment (dhimmel) [inline explore/heatmap-vega-lite/heatmap.py]: Can we switch `TOTAL` to `Total`? Gene names are CAPS by convention, but total is a regular word. (We don't want someone thinking total is a gene)

File: explore/heatmap-vega-lite/heatmap.py [hunk 2]
```diff
@@ -140,115 +141,149 @@
...
 
 # There are clearly some issues with the figure format, which proved difficult to fix at the first attempt. In order to address this problem, the changes have been made directly in the JSON file. At this intial stage, it appears reasonable to edit the JSON data in Vega-Lite editor, due to a limited functionality of Altair API (as well as the author's lack of experience with Altair). 
 # 
 # The following is the edited version JSON file produced in the previous cell. Since the main focus of this exercise is the figure formating, the actual data has been removed (See "values": [.......]).
 # * To improve the appearance of the column labels, an attribute "axis" has been added, with two sub-attributes: "labelAngle" (rotates the labels), and "offset" (shifts the labels vertically).
 # * More exploring needs to be done...
-
-# In[ ]:
-
 {
```
Comment (dhimmel) [inline explore/heatmap-vega-lite/heatmap.py]: Should this be in a markdown cell with a codeblock?

**Comment by George-Zipperlen** [comment]
> On Apr 13, 2017, at 3:19 PM, Daniel Himmelstein <notifications@github.com> wrote:
> 
> @dhimmel commented on this pull request.
> 
> In explore/heatmap-vega-lite/heatmap.py <https://github.com/cognoma/machine-learning/pull/91#discussion_r111466514>:
> 
> > @@ -123,14 +124,14 @@
>  y = y_full.assign(indicator = y_full.max(axis=1))
>  
>  
> -# In[12]:
> +# In[11]:
>  
>  unique_pos = y.groupby('disease').apply(lambda x: x['indicator'].sum())
>  heatmap_df0 = y_full.groupby('disease').sum().assign(TOTAL = unique_pos)
> Can we switch TOTAL to Total? Gene names are CAPS by convention, but total is a regular word. (We don't want someone thinking total is a gene)
> 
This is a good idea. 

To make it even more distinct, "~Total" would also force the Altair/Vega-Lite module to sort  it to the end, 
without more dataframe manipulation.

It's a bit of a kludge, but the clean solution of making a two faceted graph with the genes in one heatmap,
and a single column 'Total' heatmap in the other, will take some more work.


> 
> In explore/heatmap-vega-lite/heatmap.py <https://github.com/cognoma/machine-learning/pull/91#discussion_r111466685>:
> 
> >  
>  # There are clearly some issues with the figure format, which proved difficult to fix at the first attempt. In order to address this problem, the changes have been made directly in the JSON file. At this intial stage, it appears reasonable to edit the JSON data in Vega-Lite editor, due to a limited functionality of Altair API (as well as the author's lack of experience with Altair). 
>  # 
>  # The following is the edited version JSON file produced in the previous cell. Since the main focus of this exercise is the figure formating, the actual data has been removed (See "values": [.......]).
>  # * To improve the appearance of the column labels, an attribute "axis" has been added, with two sub-attributes: "labelAngle" (rotates the labels), and "offset" (shifts the labels vertically).
>  # * More exploring needs to be done...
> -
> -# In[ ]:
> -
>  {
> Should this be in a markdown cell with a codeblock?
> 

I should remove this comment altogether. Now that I've figured out how to do the transformation,  it is part of the Python Altair call.
The remaining low-level problem I'm tackling is to fix the orientation and labelling of the 'Legend' part of the graph.

**Comment by dhimmel** [comment]
> To make it even more distinct, "~Total" would also force the Altair/Vega-Lite module to sort  it to the end, without more dataframe manipulation.

What if you use `"\u200BTotal"` in Python? This will add an invisible unicode space. Perhaps that will fix the sort.

**Comment by George-Zipperlen** [comment]
> What if you use "\u200BTotal"? This will add an invisible unicode space. Perhaps that will fix the sort.

Doesn't work as quick fix, TOTAL is both a column name, and a Python variable name.  One of those
things I don't understand yet about pandas.

For now, using _Total which sorts after the uppercase gene symbols

**Comment by patrick-miller** [comment]
The invisible unicode space should work if you don't use the syntactic sugar and use a dictionary instead.

Instead of:
`heatmap_df0 = y_full.groupby('disease').sum().assign(TOTAL = unique_pos)`

Use:
`heatmap_df0 = y_full.groupby('disease').sum().assign(**{'\u200BTotal' : unique_pos})`

**Comment by George-Zipperlen** [comment]
> On Apr 13, 2017, at 6:18 PM, Patrick Miller <notifications@github.com> wrote:
> 
> The invisible unicode space should work if you don't use the syntactic sugar and use a dictionary instead:
> 
> heatmap_df0 = y_full.groupby('disease').sum().assign(TOTAL = unique_pos)
> 
> heatmap_df0 = y_full.groupby('disease').sum().assign(**{'\u200BTotal' : unique_pos})
> 
> 


Thank you.  That actually makes more sense to me than the syntactic sugar did.

**Comment by dhimmel** [comment]
@George-Zipperlen it looks like you opened another pull request: https://github.com/cognoma/machine-learning/pull/92. You should be not be opening new PRs, unless they're for a different addition. Instead just push more commits to the branch that the pull request is based off of.

So pick this PR #91 (this one) or #92 and stick with it.

**Comment by dhimmel** [comment]
@patrick-miller nice workaround.

You can also use the more traditional pandas syntax:

```python
heatmap_df0['\u200BTotal'] = unique_pos
```

**Comment by patrick-miller** [comment]
@dhimmel agreed, that should be preferred.

**Review by dhimmel — APPROVED** [review]
@George-Zipperlen nice job. Let me know if there's anything else you want to do before I merge.

**Comment by George-Zipperlen** [comment]
I think I've merged all the changes into this pull request: #91 (fix-heatmap-labels), sorry about 
that -- I confused branches in my local git repository with pull requests on github. 

This branch should now have the following changes to the altair/vega-lite heatmap:
- ability to change colormaps, in particular, use the viridis colormap
- move the gene-symbol labels to the bottom of the graph, labels shown vertically
- rename the 'Total' column to distinguish it from gene columns, and move it to the right
- show only the maximum and minimum data values in the legend labels, to get rid of label blob

I can't figure out how to rotate the heatmap legend from horizontal to vertical in vega-lite.  
They say that the 'Legend' configuration does for color and symbol legends what the 'Axis' 
configuration does for horizontal and vertical axes (labels, tick-marks, etc),
but the set of Legend properties is smaller than the set of Axis properties. 

@dhimmel Thank you for your patience as I learned this git/github/cognoma workflow

**Comment by George-Zipperlen** [comment]
@dhimmel : I think that's it for now, if it looks good to you

**Comment by dhimmel** [comment]
Thanks @George-Zipperlen

**Pull request merged**